File: scripts/markdown_report_contract.mjs
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

import { z } from "zod";

const PLACEHOLDER = /{{([^{}\n]+)}}/g;
const SENTENCE_END = /[.!?](?:\s|$)/g;
const PROTECTED_TOKEN = /(?:\b(?:[A-Z]{2,}(?:-[A-Z0-9]+)+|\d{4}-W\d{2}|\d{4}-\d{2}-\d{2})\b|\b\d+(?:\.\d+)?%?\b|\b(?:MYR|USD|EUR)\s*\d+(?:[.,]\d+)*\b|(?:https?:\/\/|[a-z]+:\/\/)[^\s)]+|\[[^\]]+\]\([^)]+\))/g;

const FieldBase = z.object({
  name: z.string().regex(/^[a-z][a-z0-9_]*$/),
  heading: z.string().trim().min(1),
  placeholder: z.string().trim().min(1).regex(/^[^{}\n]+$/),
  optional: z.boolean(),
  cleanup: z.boolean(),
  description: z.string().trim().min(1),
}).strict();

const ScalarField = FieldBase.extend({
  kind: z.literal("scalar"),
  sentences: z.number().int().positive().optional(),
});

const TableField = FieldBase.extend({
  kind: z.literal("table"),
  columns: z.array(z.object({
    key: z.string().regex(/^[a-z][a-z0-9_]*$/),
    heading: z.string().trim().min(1),
    description: z.string().trim().min(1).optional(),
  }).strict()).min(1),
  min_rows: z.number().int().nonnegative().optional(),
  max_rows: z.number().int().nonnegative().optional(),
}).superRefine((field, context) => {
  const keys = field.columns.map((column) => column.key);
  if (new Set(keys).size !== keys.length) {
    context.addIssue({ code: "custom", message: "column keys must be unique" });
  }
  if (field.min_rows !== undefined && field.max_rows !== undefined && field.max_rows < field.min_rows) {
    context.addIssue({ code: "custom", message: "max_rows must be greater than or equal to min_rows" });
  }
});

const InterpretationSchema = z.object({
  template_id: z.string().regex(/^[a-z0-9]+(?:-[a-z0-9]+)*$/),
  template_version: z.string().regex(/^\d+\.\d+\.\d+$/),
  fields: z.array(z.discriminatedUnion("kind", [ScalarField, TableField])).min(1),
}).strict();

function normalize(value) {
  return value.replace(/\r\n/g, "\n").replace(/[ \t]+/g, " ").replace(/\n[ \t]+/g, "\n").trim();
}

function stripGoldenExample(comment) {
  return normalize(comment.replace(/GOLDEN EXAMPLE[\s\S]*?END GOLDEN EXAMPLE/gi, ""));
}

function parseFrontmatter(markdown) {
  const match = markdown.match(/^---\n([\s\S]*?)\n---\n/);
  if (!match) throw new Error("Template must start with Markdown frontmatter.");
  const keys = match[1].split("\n").filter(Boolean).map((line) => line.match(/^([A-Za-z0-9_-]+):/)?.[1]).filter(Boolean);
  const values = Object.fromEntries(match[1].split("\n").map((line) => {
    const found = line.match(/^([A-Za-z0-9_-]+):\s*(.*)$/);
    return found ? [found[1], found[2].replace(/^['\"]|['\"]$/g, "")] : ["", ""];
  }).filter(([key]) => key));
  return { raw: match[0], keys, values, body: markdown.slice(match[0].length) };
}

export function inspectMarkdownTemplate(markdown) {
  const frontmatter = parseFrontmatter(markdown);
  const headingMatches = [...frontmatter.body.matchAll(/^## (.+)$/gm)];
  const sections = headingMatches.map((match, index) => {
    const start = match.index + match[0].length;
    const end = headingMatches[index + 1]?.index ?? frontmatter.body.length;
    const source = frontmatter.body.slice(start, end).trim();
    const comment = source.match(/^<!--([\s\S]*?)-->/)?.[1] ?? "";
    const table = source.match(/^\|(.+)\|\n\|(?:\s*:?-+)/m);
    const columns = table ? table[1].split("|").map((item) => item.trim()) : [];
    const placeholders = [...source.matchAll(PLACEHOLDER)].map((item) => item[1]);
    const golden = comment.match(/GOLDEN EXAMPLE[^\n]*\n([\s\S]*?)\nEND GOLDEN EXAMPLE/i)?.[1]?.trim() ?? null;
    return {
      heading: match[1].trim(),
      instruction: stripGoldenExample(comment),
      raw_instruction: normalize(comment),
      columns,
      placeholders: [...new Set(placeholders)],
      golden_example: golden,
    };
  });
  return {
    template_id: frontmatter.values.template_id,
    template_version: frontmatter.values.template_version,
    frontmatter_keys: frontmatter.keys,
    sections,
  };
}

function fail(issues) {
  const error = new Error(`Template interpretation is ambiguous or incompatible:\n${issues.map((issue) => `- ${issue}`).join("\n")}`);
  error.name = "TemplateContractError";
  error.issues = issues;
  throw error;
}

function scalarSchema(field) {
  let schema = z.string().trim().min(1).describe(field.description);
  if (field.sentences) {
    schema = schema.refine((value) => (value.match(SENTENCE_END) ?? []).length === field.sentences, {
      message: `${field.name} must contain exactly ${field.sentences} sentences.`,
    });
  }
  return field.optional ? schema.optional() : schema;
}

function rowSchema(field) {
  const shape = Object.fromEntries(field.columns.map((column) => [
    column.key,
    z.string().trim().min(1).describe(column.description || column.heading),
  ]));
  let schema = z.array(z.object(shape).strict()).describe(field.description);
  if (field.min_rows !== undefined) schema = schema.min(field.min_rows);
  if (field.max_rows !== undefined) schema = schema.max(field.max_rows);
  return field.optional ? schema.optional() : schema;
}

export function compileMarkdownReportContract(markdown, interpretation) {
  const observed = inspectMarkdownTemplate(markdown);
  const issues = [];
  const parsedInterpretation = InterpretationSchema.safeParse(interpretation);
  if (!parsedInterpretation.success) {
    fail([`interpretation shape is invalid: ${z.prettifyError(parsedInterpretation.error)}`]);
  }
  interpretation = parsedInterpretation.data;
  if (interpretation.template_id !== observed.template_id) issues.push(`template_id expected ${observed.template_id}, received ${interpretation.template_id}.`);
  if (interpretation.template_version !== observed.template_version) issues.push(`template_version expected ${observed.template_version}, received ${interpretation.template_version}.`);

  const observedByHeading = new Map(observed.sections.map((section) => [section.heading, section]));
  const interpretedPlaceholders = [];
  for (const field of interpretation.fields) {
    const section = observedByHeading.get(field.heading);
    interpretedPlaceholders.push(field.placeholder);
    if (!section) {
      issues.push(`field ${field.name} references missing heading "${field.heading}".`);
      continue;
    }
    if (!section.placeholders.includes(field.placeholder)) issues.push(`${field.heading}: placeholder {{${field.placeholder}}} was not found.`);
    if (section.instruction && normalize(field.description) !== section.instruction) {
      issues.push(`${field.heading}: description must copy the Markdown instruction nearly verbatim; expected "${section.instruction}".`);
    }
    if (/\bexactly\s+three\b/i.test(section.instruction) && field.sentences !== 3) issues.push(`${field.heading}: "Exactly three" requires sentences: 3.`);
    if (field.kind === "table") {
      const interpretedColumns = (field.columns ?? []).map((column) => column.heading);
      if (JSON.stringify(interpretedColumns) !== JSON.stringify(section.columns)) {
        issues.push(`${field.heading}: table columns differ; expected ${JSON.stringify(section.columns)}, received ${JSON.stringify(interpretedColumns)}.`);
      }
      if (field.min_rows === undefined && !field.optional) issues.push(`${field.heading}: table cardinality is ambiguous; supply min_rows or mark optional.`);
    }
  }
  const observedPlaceholders = observed.sections.flatMap((section) => section.placeholders);
  const duplicateObservedPlaceholders = [...new Set(observedPlaceholders.filter(
    (placeholder, index) => observedPlaceholders.indexOf(placeholder) !== index,
  ))];
  for (const placeholder of duplicateObservedPlaceholders) {
    issues.push(`body placeholder {{${placeholder}}} appears in more than one report section; placeholders must be unique.`);
  }
  for (const placeholder of observedPlaceholders) {
    const count = interpretedPlaceholders.filter((candidate) => candidate === placeholder).length;
    if (count === 0) issues.push(`placeholder {{${placeholder}}} has no interpreted field.`);
    if (count > 1) issues.push(`placeholder {{${placeholder}}} is mapped by more than one interpreted field.`);
  }
  for (const placeholder of interpretedPlaceholders) {
    if (!observedPlaceholders.includes(placeholder)) issues.push(`interpreted placeholder {{${placeholder}}} does not exist in the report body.`);
  }
  const names = interpretation.fields.map((field) => field.name);
  if (new Set(names).size !== names.length) issues.push("Interpreted field names must be unique.");
  if (issues.length) fail(issues);

  const shape = Object.fromEntries(interpretation.fields.map((field) => [
    field.name,
    field.kind === "table" ? rowSchema(field) : scalarSchema(field),
  ]));
  const schema = z.object(shape).strict().describe(
    `Structured extraction for ${observed.template_id}@${observed.template_version}; compiled deterministically from a reviewed Markdown interpretation.`,
  );
  return { observed, interpretation: structuredClone(interpretation), schema, json_schema: z.toJSONSchema(schema, { target: "draft-2020-12" }) };
}

function protectedTokens(value) {
  return [...value.matchAll(PROTECTED_TOKEN)].map((match) => match[0].replace(/[.,;:!?]+$/, "")).sort();
}

export function applyConstrainedProseCleanup(data, replacements, interpretation) {
  const output = structuredClone(data);
  const fieldByName = new Map(interpretation.fields.map((field) => [field.name, field]));
  for (const [path, replacement] of Object.entries(replacements)) {
    const [fieldName, rowText, columnKey] = path.split(".");
    const field = fieldByName.get(fieldName);
    if (!field?.cleanup) fail([`${path}: field is not approved for prose cleanup.`]);
    const rowIndex = rowText === undefined ? null : Number(rowText);
    const current = rowIndex === null ? output[fieldName] : output[fieldName]?.[rowIndex]?.[columnKey];
    if (typeof current !== "string" || typeof replacement !== "string") fail([`${path}: cleanup target and replacement must both be strings.`]);
    const before = protectedTokens(current);
    const after = protectedTokens(replacement);
    if (JSON.stringify(before) !== JSON.stringify(after)) {
      fail([`${path}: cleanup changed protected facts, IDs, evidence, numbers, dates, money, or routes; before ${JSON.stringify(before)}, after ${JSON.stringify(after)}.`]);
    }
    if (rowIndex === null) output[fieldName] = replacement;
    else output[fieldName][rowIndex][columnKey] = replacement;
  }
  return output;
}

function renderTable(field, rows) {
  const cell = (value) => value.replace(/\r?\n/g, "<br>").replace(/\|/g, "\\|");
  return rows.map((row) => `| ${field.columns.map((column) => cell(row[column.key])).join(" | ")} |`).join("\n");
}

export function renderMarkdownReport(markdown, data, interpretation, frontmatterValues = {}) {
  const compiled = compileMarkdownReportContract(markdown, interpretation);
  const parsed = compiled.schema.safeParse(data);
  if (!parsed.success) throw new Error(`Structured report failed compiled Zod validation:\n${z.prettifyError(parsed.error)}`);
  const replacements = { ...frontmatterValues };
  for (const field of interpretation.fields) {
    const value = parsed.data[field.name];
    replacements[field.placeholder] = field.kind === "table" ? renderTable(field, value ?? []) : value ?? "";
  }
  const rendered = markdown.replace(PLACEHOLDER, (original, name) => Object.hasOwn(replacements, name) ? String(replacements[name]) : original);
  const unresolved = [...rendered.matchAll(PLACEHOLDER)].map((match) => match[1]);
  if (unresolved.length) throw new Error(`Rendered report has unresolved placeholders: ${[...new Set(unresolved)].join(", ")}.`);
  const renderedShape = inspectMarkdownTemplate(rendered);
  const templateShape = inspectMarkdownTemplate(markdown);
  if (JSON.stringify(renderedShape.frontmatter_keys) !== JSON.stringify(templateShape.frontmatter_keys)
      || JSON.stringify(renderedShape.sections.map(({ heading, columns }) => ({ heading, columns }))) !== JSON.stringify(templateShape.sections.map(({ heading, columns }) => ({ heading, columns })))) {
    throw new Error("Rendered report changed the Markdown template shape.");
  }
  return rendered;
}

export function diffMarkdownReportContract(markdown, interpretation) {
  try {
    const compiled = compileMarkdownReportContract(markdown, interpretation);
    return { compatible: true, template_id: compiled.observed.template_id, template_version: compiled.observed.template_version, issues: [] };
  } catch (error) {
    if (error.name !== "TemplateContractError") throw error;
    return { compatible: false, issues: error.issues };
  }
}

async function main(argv) {
  const [templatePath, interpretationPath, extractionPath] = argv;
  if (!templatePath || !interpretationPath) {
    console.error("Usage: node scripts/markdown_report_contract.mjs TEMPLATE.md INTERPRETATION.json [EXTRACTION.json]");
    process.exitCode = 2;
    return;
  }
  const markdown = readFileSync(templatePath, "utf8");
  const interpretation = JSON.parse(readFileSync(interpretationPath, "utf8"));
  const result = diffMarkdownReportContract(markdown, interpretation);
  if (result.compatible && extractionPath) {
    const compiled = compileMarkdownReportContract(markdown, interpretation);
    const parsed = compiled.schema.safeParse(JSON.parse(readFileSync(extractionPath, "utf8")));
    result.extraction_valid = parsed.success;
    if (!parsed.success) result.extraction_errors = z.treeifyError(parsed.error);
  }
  console.log(JSON.stringify(result, null, 2));
  if (!result.compatible || result.extraction_valid === false) process.exitCode = 1;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) await main(process.argv.slice(2));

File: scripts/markdown_report_contract.py
#!/usr/bin/env python3
"""Compile reviewed Markdown report interpretations into strict Pydantic models."""

from __future__ import annotations

import argparse
import copy
import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, create_model, model_validator


PLACEHOLDER = re.compile(r"{{([^{}\n]+)}}")
SENTENCE_END = re.compile(r"[.!?](?:\s|$)")
PROTECTED_TOKEN = re.compile(r"(?:\b(?:[A-Z]{2,}(?:-[A-Z0-9]+)+|\d{4}-W\d{2}|\d{4}-\d{2}-\d{2})\b|\b\d+(?:\.\d+)?%?\b|\b(?:MYR|USD|EUR)\s*\d+(?:[.,]\d+)*\b|(?:https?://|[a-z]+://)[^\s)]+|\[[^\]]+\]\([^)]+\))")
SafeName = Annotated[str, StringConstraints(pattern=r"^[a-z][a-z0-9_]*$")]
NonEmpty = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Column(StrictModel):
    key: SafeName
    heading: NonEmpty
    description: NonEmpty | None = None


class FieldInterpretation(StrictModel):
    name: SafeName
    heading: NonEmpty
    placeholder: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, pattern=r"^[^{}\n]+$")]
    kind: Literal["scalar", "table"]
    optional: bool
    cleanup: bool
    description: NonEmpty
    sentences: Annotated[int, Field(gt=0)] | None = None
    columns: list[Column] | None = None
    min_rows: Annotated[int, Field(ge=0)] | None = None
    max_rows: Annotated[int, Field(ge=0)] | None = None

    @model_validator(mode="after")
    def validate_kind(self) -> "FieldInterpretation":
        if self.kind == "table":
            if not self.columns:
                raise ValueError("table fields require columns")
            keys = [column.key for column in self.columns]
            if len(set(keys)) != len(keys):
                raise ValueError("column keys must be unique")
            if self.min_rows is not None and self.max_rows is not None and self.max_rows < self.min_rows:
                raise ValueError("max_rows must be greater than or equal to min_rows")
        elif self.columns is not None or self.min_rows is not None or self.max_rows is not None:
            raise ValueError("scalar fields forbid table settings")
        return self


class Interpretation(StrictModel):
    template_id: Annotated[str, StringConstraints(pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")]
    template_version: Annotated[str, StringConstraints(pattern=r"^\d+\.\d+\.\d+$")]
    fields: Annotated[list[FieldInterpretation], Field(min_length=1)]


class TemplateContractError(ValueError):
    def __init__(self, issues: list[str]):
        self.issues = issues
        super().__init__("Template interpretation is ambiguous or incompatible:\n" + "\n".join(f"- {issue}" for issue in issues))


@dataclass(frozen=True)
class CompiledReportContract:
    observed: dict[str, Any]
    interpretation: dict[str, Any]
    schema: type[BaseModel]
    json_schema: dict[str, Any]


def _normalize(value: str) -> str:
    return re.sub(r"\n[ \t]+", "\n", re.sub(r"[ \t]+", " ", value.replace("\r\n", "\n"))).strip()


def _frontmatter(markdown: str) -> tuple[list[str], dict[str, str], str]:
    match = re.match(r"^---\n([\s\S]*?)\n---\n", markdown)
    if not match:
        raise ValueError("Template must start with Markdown frontmatter.")
    rows = [re.match(r"^([A-Za-z0-9_-]+):\s*(.*)$", line) for line in match.group(1).splitlines()]
    pairs = [(row.group(1), row.group(2).strip("'\"")) for row in rows if row]
    return [key for key, _ in pairs], dict(pairs), markdown[match.end():]


def inspect_markdown_template(markdown: str) -> dict[str, Any]:
    keys, values, body = _frontmatter(markdown)
    headings = list(re.finditer(r"^## (.+)$", body, re.MULTILINE))
    sections = []
    for index, heading in enumerate(headings):
        source = body[heading.end() : headings[index + 1].start() if index + 1 < len(headings) else len(body)].strip()
        comment_match = re.match(r"^<!--([\s\S]*?)-->", source)
        comment = comment_match.group(1) if comment_match else ""
        golden_match = re.search(r"GOLDEN EXAMPLE[^\n]*\n([\s\S]*?)\nEND GOLDEN EXAMPLE", comment, re.IGNORECASE)
        instruction = re.sub(r"GOLDEN EXAMPLE[\s\S]*?END GOLDEN EXAMPLE", "", comment, flags=re.IGNORECASE)
        table = re.search(r"^\|(.+)\|\n\|(?:\s*:?-+)", source, re.MULTILINE)
        sections.append({
            "heading": heading.group(1).strip(),
            "instruction": _normalize(instruction),
            "raw_instruction": _normalize(comment),
            "columns": [cell.strip() for cell in table.group(1).split("|")] if table else [],
            "placeholders": list(dict.fromkeys(PLACEHOLDER.findall(source))),
            "golden_example": golden_match.group(1).strip() if golden_match else None,
        })
    return {"template_id": values.get("template_id"), "template_version": values.get("template_version"), "frontmatter_keys": keys, "sections": sections}


def build_report_model(interpretation_value: dict[str, Any], *, model_name: str = "GeneratedReport") -> type[BaseModel]:
    interpretation = Interpretation.model_validate(interpretation_value)
    definitions: dict[str, tuple[Any, Any]] = {}
    for field in interpretation.fields:
        if field.kind == "scalar":
            annotation: Any = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1), Field(description=field.description)]
        else:
            row = create_model(
                f"{model_name}{''.join(part.title() for part in field.name.split('_'))}Row",
                __base__=StrictModel,
                **{
                    column.key: (Annotated[str, StringConstraints(strip_whitespace=True, min_length=1), Field(description=column.description or column.heading)], ...)
                    for column in field.columns or []
                },
            )
            annotation = Annotated[list[row], Field(min_length=field.min_rows, max_length=field.max_rows, description=field.description)]
        if field.optional:
            annotation = annotation | None
            default: Any = None
        else:
            default = ...
        definitions[field.name] = (annotation, default)

    def validate_sentences(self: BaseModel) -> BaseModel:
        for field in interpretation.fields:
            if field.sentences and len(SENTENCE_END.findall(getattr(self, field.name))) != field.sentences:
                raise ValueError(f"{field.name} must contain exactly {field.sentences} sentences.")
        return self

    return create_model(
        model_name,
        __base__=StrictModel,
        __validators__={"validate_sentences": model_validator(mode="after")(validate_sentences)},
        **definitions,
    )


def compile_markdown_report_contract(markdown: str, interpretation_value: dict[str, Any]) -> CompiledReportContract:
    observed = inspect_markdown_template(markdown)
    try:
        interpretation = Interpretation.model_validate(interpretation_value)
    except Exception as error:
        raise TemplateContractError([f"interpretation shape is invalid: {error}"]) from error
    issues = []
    if interpretation.template_id != observed["template_id"]:
        issues.append(f"template_id expected {observed['template_id']}, received {interpretation.template_id}.")
    if interpretation.template_version != observed["template_version"]:
        issues.append(f"template_version expected {observed['template_version']}, received {interpretation.template_version}.")
    sections = {section["heading"]: section for section in observed["sections"]}
    interpreted_placeholders = []
    for field in interpretation.fields:
        interpreted_placeholders.append(field.placeholder)
        section = sections.get(field.heading)
        if not section:
            issues.append(f"field {field.name} references missing heading \"{field.heading}\".")
            continue
        if field.placeholder not in section["placeholders"]:
            issues.append(f"{field.heading}: placeholder {{{{{field.placeholder}}}}} was not found.")
        if section["instruction"] and _normalize(field.description) != section["instruction"]:
            issues.append(f"{field.heading}: description must copy the Markdown instruction nearly verbatim; expected \"{section['instruction']}\".")
        if re.search(r"\bexactly\s+three\b", section["instruction"], re.IGNORECASE) and field.sentences != 3:
            issues.append(f"{field.heading}: \"Exactly three\" requires sentences: 3.")
        if field.kind == "table":
            headings = [column.heading for column in field.columns or []]
            if headings != section["columns"]:
                issues.append(f"{field.heading}: table columns differ; expected {section['columns']}, received {headings}.")
            if field.min_rows is None and not field.optional:
                issues.append(f"{field.heading}: table cardinality is ambiguous; supply min_rows or mark optional.")
    observed_placeholders = [placeholder for section in observed["sections"] for placeholder in section["placeholders"]]
    for placeholder in sorted({value for value in observed_placeholders if observed_placeholders.count(value) > 1}):
        issues.append(f"body placeholder {{{{{placeholder}}}}} appears in more than one report section; placeholders must be unique.")
    for placeholder in observed_placeholders:
        count = interpreted_placeholders.count(placeholder)
        if count == 0:
            issues.append(f"placeholder {{{{{placeholder}}}}} has no interpreted field.")
        elif count > 1:
            issues.append(f"placeholder {{{{{placeholder}}}}} is mapped by more than one interpreted field.")
    if len({field.name for field in interpretation.fields}) != len(interpretation.fields):
        issues.append("Interpreted field names must be unique.")
    if issues:
        raise TemplateContractError(issues)
    model = build_report_model(interpretation.model_dump(mode="json"), model_name="CompiledReport")
    schema = model.model_json_schema()
    schema["description"] = f"Structured extraction for {interpretation.template_id}@{interpretation.template_version}; compiled deterministically from a reviewed Markdown interpretation."
    return CompiledReportContract(observed, interpretation.model_dump(mode="json"), model, schema)


def _protected_tokens(value: str) -> list[str]:
    return sorted(match.group(0).rstrip(".,;:!?") for match in PROTECTED_TOKEN.finditer(value))


def apply_constrained_prose_cleanup(data: dict[str, Any], replacements: dict[str, str], interpretation_value: dict[str, Any]) -> dict[str, Any]:
    output = copy.deepcopy(data)
    fields = {field.name: field for field in Interpretation.model_validate(interpretation_value).fields}
    for path, replacement in replacements.items():
        parts = path.split(".")
        field = fields.get(parts[0])
        if not field or not field.cleanup:
            raise TemplateContractError([f"{path}: field is not approved for prose cleanup."])
        current = output[parts[0]] if len(parts) == 1 else output[parts[0]][int(parts[1])][parts[2]]
        if _protected_tokens(current) != _protected_tokens(replacement):
            raise TemplateContractError([f"{path}: cleanup changed protected facts, IDs, evidence, numbers, dates, money, or routes; before {_protected_tokens(current)}, after {_protected_tokens(replacement)}."])
        if len(parts) == 1:
            output[parts[0]] = replacement
        else:
            output[parts[0]][int(parts[1])][parts[2]] = replacement
    return output


def render_markdown_report(markdown: str, data: dict[str, Any], interpretation_value: dict[str, Any], frontmatter_values: dict[str, Any] | None = None) -> str:
    compiled = compile_markdown_report_contract(markdown, interpretation_value)
    parsed = compiled.schema.model_validate(data).model_dump(mode="json")
    replacements = dict(frontmatter_values or {})
    for field in Interpretation.model_validate(interpretation_value).fields:
        value = parsed.get(field.name)
        if field.kind == "table":
            rows = []
            for row in value or []:
                cells = [str(row[column.key]).replace("\r\n", "<br>").replace("\n", "<br>").replace("|", r"\|") for column in field.columns or []]
                rows.append(f"| {' | '.join(cells)} |")
            replacements[field.placeholder] = "\n".join(rows)
        else:
            replacements[field.placeholder] = value or ""
    rendered = PLACEHOLDER.sub(lambda match: str(replacements.get(match.group(1), match.group(0))), markdown)
    unresolved = sorted(set(PLACEHOLDER.findall(rendered)))
    if unresolved:
        raise ValueError(f"Rendered report has unresolved placeholders: {', '.join(unresolved)}.")
    before, after = inspect_markdown_template(markdown), inspect_markdown_template(rendered)
    before_shape = [(section["heading"], section["columns"]) for section in before["sections"]]
    after_shape = [(section["heading"], section["columns"]) for section in after["sections"]]
    if before["frontmatter_keys"] != after["frontmatter_keys"] or before_shape != after_shape:
        raise ValueError("Rendered report changed the Markdown template shape.")
    return rendered


def diff_markdown_report_contract(markdown: str, interpretation: dict[str, Any]) -> dict[str, Any]:
    try:
        compiled = compile_markdown_report_contract(markdown, interpretation)
        return {"compatible": True, "template_id": compiled.observed["template_id"], "template_version": compiled.observed["template_version"], "issues": []}
    except TemplateContractError as error:
        return {"compatible": False, "issues": error.issues}


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("template", type=Path)
    parser.add_argument("interpretation", type=Path)
    parser.add_argument("extraction", type=Path, nargs="?")
    args = parser.parse_args()
    markdown = args.template.read_text(encoding="utf-8")
    interpretation = json.loads(args.interpretation.read_text(encoding="utf-8"))
    result = diff_markdown_report_contract(markdown, interpretation)
    if result["compatible"] and args.extraction:
        try:
            compile_markdown_report_contract(markdown, interpretation).schema.model_validate_json(args.extraction.read_text())
            result["extraction_valid"] = True
        except Exception as error:
            result["extraction_valid"] = False
            result["extraction_errors"] = str(error)
    print(json.dumps(result, indent=2))
    return 0 if result["compatible"] and result.get("extraction_valid", True) else 1


if __name__ == "__main__":
    raise SystemExit(main())
